### PR TITLE
feat: identify DOM events using CSS locator

### PR DESCRIPTION
**Pull request by limhjgrace**
Users currently configure the DOM event plugin to listen to events either with (1) the ID of the DOM element or (2) the actual JavaScript object. However, a customer may want to add a tag to certain elements which should be recorded. There is currently no way to listen for a certain tag.

This feature allows customers to specify listeners for DOM elements using CSS locators.

---

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/plugins/event-plugins/DomEventPlugin.ts [hunk 1]
```diff
@@ -84,12 +89,13 @@ export class DomEventPlugin implements Plugin {
...
 
-    private getEventListener(): EventListener {
+    private getEventListener(domEvent?: TargetDomEvent): EventListener {
         return (event: Event): void => {
             const eventData: DomEvent = {
                 version: '1.0.0',
                 event: event.type,
-                elementId: this.getElementId(event)
+                elementId: this.getElementId(event),
+                cssLocator: this.getElementCSSLocator(event, domEvent)
```
Comment (qhanam) [inline src/plugins/event-plugins/DomEventPlugin.ts:98]: Since the schema specifies that `cssLocator` is type `string`, we omit the `cssLocator` field to the event object if `cssLocator` is undefined.

File: src/event-schemas/dom-event.json [hunk 2]
```diff
@@ -16,8 +16,12 @@
...
             "description": "DOM element ID."
+        },
+        "cssLocator": {
+            "type": "string",
+            "description": "CSS Locator string."
         }
     },
     "additionalProperties": false,
-    "required": ["version", "event", "elementId"]
+    "required": ["version", "event", "elementId", "cssLocator"]
```
Comment (qhanam) [inline src/event-schemas/dom-event.json:26]: We probably want `cssLocator` field to be an optional field. We always provide `elementId`, but we do not always provide `cssLocator`.

File: src/plugins/event-plugins/DomEventPlugin.ts [hunk 3]
```diff
@@ -113,15 +119,33 @@ export class DomEventPlugin implements Plugin {
...
+        }
+
+        if (event.target instanceof Element && domEvent.cssLocator) {
+            return domEvent.cssLocator;
+        }
+
+        if (event.target instanceof Node) {
+            return event.target.nodeName;
+        }
+    }
```
Comment (qhanam) [inline src/plugins/event-plugins/DomEventPlugin.ts:134]: Unlike the `elementId`, we can consider the CSS locator optional. In this case we can return a CSS locator if one exists, or `undefined` otherwise.
```javascript
private getElementCSSLocator(domEvent: TargetDomEvent) {
    if (domEvent.cssLocator) {
        return domEvent.cssLocator;
    }
    return undefined;
}
```

File: src/plugins/event-plugins/DomEventPlugin.ts [hunk 4]
```diff
@@ -113,15 +119,33 @@ export class DomEventPlugin implements Plugin {
...
         const eventType = domEvent.event;
-        const eventListener = this.getEventListener();
+        const eventListener = this.getEventListener(domEvent);
         this.eventListenerMap.set(domEvent, eventListener);
 
         if (domEvent.elementId) {
             document
                 .getElementById(domEvent.elementId)
                 ?.addEventListener(eventType, eventListener);
+        } else if (domEvent.cssLocator) {
```
Comment (qhanam) [inline src/plugins/event-plugins/DomEventPlugin.ts:144]: If there are multiple locators in `domEvent`, which locator should be used? I.e., right now it is `elementID -> cssLocator -> element`. We could also throw an error, but I think it's better to have a sane default here instead.

I think we should add a unit test that defines which locator is used in this case.

File: src/event-schemas/dom-event.json (hunk 2 above)
Comment (limhjgrace) [reply src/event-schemas/dom-event.json:26]: Removed `cssLocator` from being a required property.

File: src/plugins/event-plugins/DomEventPlugin.ts (hunk 1 above)
Comment (limhjgrace) [reply src/plugins/event-plugins/DomEventPlugin.ts:98]: Refactored so that `cssLocator` field is only added when defined.

File: src/plugins/event-plugins/DomEventPlugin.ts (hunk 3 above)
Comment (limhjgrace) [reply src/plugins/event-plugins/DomEventPlugin.ts:134]: Refactored as suggested.

File: src/plugins/event-plugins/DomEventPlugin.ts (hunk 4 above)
Comment (limhjgrace) [reply src/plugins/event-plugins/DomEventPlugin.ts:144]: As discussed, we'll add listeners for all fields that are passed in.

File: src/plugins/event-plugins/DomEventPlugin.ts [hunk 5]
```diff
@@ -36,13 +41,18 @@ const defaultConfig: DomEventPluginConfig = {
 export class DomEventPlugin implements Plugin {
     private recordEvent: RecordEvent | undefined;
     private pluginId: string;
-    private eventListenerMap: Map<TargetDomEvent, EventListener>;
+    private eventListenerMap: Map<TargetDomEvent, Map<string, EventListener>>;
```
Comment (qhanam) [inline src/plugins/event-plugins/DomEventPlugin.ts:44]: **suggestion:** The problem we have here is that the TargetDomEvent can potentially contain three selectors: ID, CSS locator or object. Having a map within a map makes things a little hard to follow. Consider a different strategy:

1. Map a `TargetDomEvent` to an array of `EventListener`. There will be either one or two items in the array. This will remove a lot of the complexity, because we don't need to think about the 'type' of event listener.

2. Only use one of the three selectors. This adds the least complexity. We can document that only one of selectors should be given, and the order in which the selectors will be preferred.

File: src/plugins/event-plugins/DomEventPlugin.ts [hunk 6]
```diff
@@ -84,13 +94,16 @@ export class DomEventPlugin implements Plugin {
         );
     }
 
-    private getEventListener(): EventListener {
+    private getEventListener(domEvent?: TargetDomEvent): EventListener {
```
Comment (qhanam) [inline src/plugins/event-plugins/DomEventPlugin.ts:97]: **nit:** We can make this a bit more readable by changing the `domEvent` parameter to `cssLocator`

File: src/plugins/event-plugins/DomEventPlugin.ts [hunk 7]
```diff
@@ -113,32 +126,82 @@ export class DomEventPlugin implements Plugin {
...
+        // add event listener to element identified by the element ID if no CSS locator provided or
+        // the element is not one of the elements identified by the CSS locator
+        if (domEvent.elementId && !isElementIdentifiedByLocator) {
             document
                 .getElementById(domEvent.elementId)
-                ?.addEventListener(eventType, eventListener);
+                ?.addEventListener(eventType, eventListenerID);
+
+            // safe default
         } else if (domEvent.element) {
```
Comment (qhanam) [inline src/plugins/event-plugins/DomEventPlugin.ts:165]: **issue:** If we are adding event listeners to **all** selectors (i.e., ID, object and CSS) in a given `TargetDomEvent`, this should be an `if` statement rather than an `else if` statement.

File: src/plugins/event-plugins/DomEventPlugin.ts (hunk 6 above)
Comment (limhjgrace) [reply src/plugins/event-plugins/DomEventPlugin.ts:97]: Refactored

File: src/plugins/event-plugins/DomEventPlugin.ts [hunk 8]
```diff
@@ -84,13 +94,16 @@ export class DomEventPlugin implements Plugin {
...
-    private getEventListener(): EventListener {
+    private getEventListener(domEvent?: TargetDomEvent): EventListener {
         return (event: Event): void => {
             const eventData: DomEvent = {
                 version: '1.0.0',
                 event: event.type,
                 elementId: this.getElementId(event)
             };
+            if (domEvent?.cssLocator) {
+                eventData.cssLocator = this.getElementCSSLocator(domEvent);
```
[the remark this answers is not in the document]
Comment (limhjgrace) [reply src/plugins/event-plugins/DomEventPlugin.ts:105]: Did you mean:
if (cssLocator !== undefined) {
    eventData.cssLocator = cssLocator;

File: src/plugins/event-plugins/DomEventPlugin.ts (hunk 7 above)
Comment (limhjgrace) [reply src/plugins/event-plugins/DomEventPlugin.ts:165]: Refactored

File: src/plugins/event-plugins/DomEventPlugin.ts (hunk 5 above)
Comment (limhjgrace) [reply src/plugins/event-plugins/DomEventPlugin.ts:44]: Will go with option 2. Preference will be: CSS Selector > Element ID > Element. CSS Selector will be preferred if provided because it point to the most elements while element ID points to only one element.

**Review by qhanam — APPROVED** [review]
